content/blog/2020-04-dplyr-1-0-0-for-package-developers/index.Rmarkdown

```diff
@@ -0,0 +1,120 @@
+slug: dplyr-1-0-0-package-dev
+description: > 
+  dplyr 1.0.0 is scheduled for release on May 15. This blog post talks
+  about what package maintainers can do to prepare.
+categories:
+  - package
+tags:
+  - dplyr
+photo:
+  author: Tekton
+  url: https://unsplash.com/photos/LtphNTXHQAc
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
+```
+
+As you're hopefully aware, [dplyr 1.0.0 is coming soon](https://www.tidyverse.org/blog/2020/03/dplyr-1-0-0-is-coming-soon/), and we've been writing a [series of blog posts](https://www.tidyverse.org/tags/dplyr/) about the user-facing changes that you, as a data scientist have to look forward to. Today, I wanted to change tack a little and talk about the changes from the perspective of the package developer.
+
+But first, an update on the release process: in the process of preparing for the release, we discovered some subtle problems that arise when combining different types of data frames (including [data.table](http://r-datatable.com)s). It took us a little while to figure out what we (and package developers need to do), so we've decided to push back the dplyr release: we're now planning on releasing dplyr 1.0.0 to CRAN on May 15. We're sorry that its going to longer than expected, but this gives package authors who use dplyr more time to handle changes.
+
+In this post, I want to talk a little bit about how dplyr changes might break package code, then discuss some of the major pain points a package developer might experience, and how to get help if you need it.
+
+```{r setup}
+library(dplyr, warn.conflicts = FALSE)
+```
+
+## Breaking changes
+
+There are three main ways an update to a package might break your existing code:
+
+* We've introduced a bug. Obviously, we do our best to make sure this doesn't 
+  happen (by using software development best practices like unit testing and 
+  code review) but it's impossible to eliminate all bugs.
+  
+* We've fixed a bug or otherwise made change we think is harmless. Sometimes 
+  your code accidentally depends on a behaviour that we think is incorrect and 
+  we change it. The change will be an improvement for most people, but 
+  unfortunately it [breaks your code](https://xkcd.com/1172/).
+  
+* We've deliberately made a backward incompatible interface change. We 
+  occasionally make these changes and always to significantly improve usability 
+  or consistency. Unless the package or function is experimental, we do our best 
+  to make such changes gradually, so that there's a deprecation period before 
+  the behaviour goes away altogether.
+
+dplyr 1.0.0 contains very few backward incompatible changes, but it does make a large number of changes that we believe are mostly harmless or minor improvements.  The vast majority of these will not affect data analysis code, but can affect package code, particularly unit tests. To give you a flavour for what I mean here, dplyr now preserves the names of atomic vectors:
```

Comment:
> The vast majority of these will not affect data analysis code, but can affect package code

This implies the vast majority can affect package code. If intended: good. If not, perhaps "but some can…" or something of the like.

Reply:
> The vast majority of these will not affect data analysis code, but some can affect packages, particularly through their unit tests.

Reply:
I like it